README.md
# WG0-pipeline-Preprocessing
IMPORTANT: this pipeline is in development and not yet ready for use

CHANGELOG.md
# Changelog

All notable changes implemented in this branch compared to the main branch of [WG0-pipeline-preprocessing](https://github.com/sc-eQTLgen-consortium/WG0-pipeline-preprocessing) (TBD, SHA: TBD) are documented in this file. 

Note that this branch is in beta and version 1.0.0 is not yet ready for release.

## [2.0.0] - Preprocessing - YYYY-mm-dd

#### Additions

#### Fixes

#### Changes


scripts/combine_results.py
#!/usr/bin/env python
import argparse
import os

parser = argparse.ArgumentParser(description="")
parser.add_argument("--pools", required=True, nargs="+", type=str, help="")
parser.add_argument("--counts", required=True, nargs="+", type=str, help="")
parser.add_argument("--barcodes", required=True, nargs="+", type=str, help="")
parser.add_argument("--bam", required=True, nargs="+", type=str, help="")
parser.add_argument("--out", required=True, type=str, help="The output directory where results will be saved.")
args = parser.parse_args()

if not os.path.isdir(args.out):
    os.makedirs(args.out, exist_ok=True)

print("Options in effect:")
arguments = {}
for arg in vars(args):
    print("  --{} {}".format(arg, getattr(args, arg)))
    arguments[arg] = getattr(args, arg)
print("")

import pandas as pd

# Construct the file_directories.tsv file that points to the files we will use for WG1.
file_directories = pd.DataFrame({
    'Pool': args.pools,
    'Counts': args.counts,
    'Barcodes': args.barcodes,
    'Bam': args.bam,
})
print("Saving file directories for WG1.")
file_directories.to_csv(os.path.join(args.out, "wg0_file_directories.tsv"), sep="\t", index=False)


scripts/plot_cellbender.py
#!/usr/bin/env python
# Adapted from https://github.com/broadinstitute/CellBender/blob/master/cellbender/remove_background/report.py
import argparse
import os

parser = argparse.ArgumentParser(description="")
parser.add_argument("--samples", required=True, nargs="+", type=str, help="")
parser.add_argument("--input_h5", required=True, nargs="+", type=str, help="")
parser.add_argument("--settings", required=True, nargs="+", type=str, help="")
parser.add_argument("--raw_h5", required=True, nargs="+", type=str, help="")
parser.add_argument("--max_plots_per_page", required=False, default=5, type=int, help="")
parser.add_argument("--out", required=True, type=str, help="The output directory where results will be saved.")
args = parser.parse_args()

if not os.path.isdir(args.out):
    os.makedirs(args.out, exist_ok=True)

from cellbender.remove_background.downstream import load_anndata_from_input, load_anndata_from_input_and_output
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import json
import scipy.stats
import torch


settings_info = {
    "use_cuda": ("cuda", None),
    "expected_cells": ("expected_cell_count", None),
    "total_droplets_included": ("total_droplets", None),
    "force_cell_umi_prior": ("force_cell_umi_prior", None),
    "force_empty_umi_prior": ("force_empty_umi_prior", None),
    "model": ("model", "full"),
    "epochs": ("epochs", 150),
    "low_count_threshold": ("low_count_threshold", 5),
    "z_dim":( "z_hidden_dims", 64),
    "z_layers": ("z_layers", [512]),
    "training_fraction": ("training_fraction", 0.9),
    "empty_drop_training_fraction": ("fraction_empties", 0.2),
    "ignore_features": ("blacklisted_genes", []),
    "fpr": ("fpr", [0.01]),
    "exclude_feature_types": ("exclude_features", []),
    "projected_ambient_count_threshold": ("ambient_counts_in_cells_low_limit", 0.1),
    "learning_rate": ("learning_rate", 1e-4),
    "checkpoint_mins": ("checkpoint_min", 7.),
    "final_elbo_fail_fraction": ("final_elbo_fail_fraction", None),
    "epoch_elbo_fail_fraction": ("epoch_elbo_fail_fraction", None),
    "num_training_tries": ("num_training_tries", 1),
    "learning_rate_retry_mult": ("learning_rate_retry_mult", 0.2),
    "posterior_batch_size": ("posterior_batch_size", 128),
    # "posterior_regularization": ("posterior_regularization", None),
    "alpha": ("prq_alpha", None),
    "q": ("cdf_threshold_q", None),
    # "estimator": ("estimator", None),
    "estimator_multiple_cpu": ("use_multiprocessing_estimation", False),
    "constant_learning_rate": ("constant_learning_rate", False),
    "debug": ("debug", False)
}

def assess_overall_count_removal(adata):
    cells = (adata.obs['cell_probability'] > 0.5)
    initial_counts = adata.layers['raw'][cells].sum()
    removed_counts = initial_counts - adata.layers['cellbender'][cells].sum()
    removed_percentage = removed_counts / initial_counts * 100

    estimated_ambient_per_droplet = np.exp(adata.uns['empty_droplet_size_lognormal_loc']).item()
    expected_fraction_removed_from_cells = estimated_ambient_per_droplet * cells.sum() / initial_counts

    fpr = adata.uns['target_false_positive_rate'].item()  # this is an np.ndarray with one element
    cohort_mode = False
    if type(fpr) != float:
        cohort_mode = True

    if not cohort_mode:
        expected_percentage = (expected_fraction_removed_from_cells + fpr) * 100
    else:
        expected_percentage = expected_fraction_removed_from_cells * 100

    if np.abs(expected_percentage - removed_percentage) <= 0.5:
        return {"overall_count_removal": "great"}
    elif np.abs(expected_percentage - removed_percentage) <= 1:
        return {"overall_count_removal": "decent"}
    elif np.abs(expected_percentage - removed_percentage) <= 5:
        if removed_percentage < expected_percentage:
            return {"overall_count_removal": "bit less"}
        else:
            return {"overall_count_removal": "bit more"}
    elif removed_percentage - expected_percentage > 5:
        return {"overall_count_removal": "more"}
    elif expected_percentage - removed_percentage > 5:
        return {"overall_count_removal": "fewer"}

def assess_count_removal_per_gene(adata,
                                  raw_full_adata,
                                  input_layer_key='raw',
                                  r_squared_cutoff=0.5):
    # how well does it correlate with our expectation about the ambient RNA profile?
    cells = (adata.obs['cell_probability'] > 0.5)
    counts = np.array(raw_full_adata.X.sum(axis=1)).squeeze()
    clims = [adata.obs[f'n_{input_layer_key}'][~cells].mean() / 2,
             np.percentile(adata.obs[f'n_{input_layer_key}'][cells].values, q=2)]
    # if all are called "cells" then clims[0] will be a nan
    if np.isnan(clims[0]):
        clims[0] = counts.min()
    if 'approximate_ambient_profile' in adata.uns.keys():
        approximate_ambient_profile = adata.uns['approximate_ambient_profile']
    else:
        empty_count_matrix = raw_full_adata[(counts > clims[0]) & (counts < clims[1])].X
        if empty_count_matrix.shape[0] > 100:
            approximate_ambient_profile = np.array(raw_full_adata[(counts > clims[0])
                                                                  & (counts < clims[1])].X.mean(axis=0)).squeeze()
        else:
            # a very rare edge case I've seen once
            approximate_ambient_profile = np.array(raw_full_adata[counts < clims[1]].X.mean(axis=0)).squeeze()
        approximate_ambient_profile = approximate_ambient_profile / approximate_ambient_profile.sum()
    y = adata.var['n_removed'] / adata.var['n_removed'].sum()

    cutoff = 1e-6
    logic = np.logical_not((approximate_ambient_profile < cutoff) | (y < cutoff))
    r_squared_result = scipy.stats.pearsonr(np.log(approximate_ambient_profile[logic]),
                                            np.log(y[logic]))
    if hasattr(r_squared_result, 'statistic'):
        # scipy version 1.9.0+
        r_squared = r_squared_result.statistic
    else:
        r_squared = r_squared_result[0]

    if r_squared > r_squared_cutoff:
        return {"per_gene_removal": "correct"}
    else:
        return {"per_gene_removal": "incorrect"}

def assess_learning_curve(adata,
                          spike_size: float = 0.5,
                          deviation_size: float = 0.25,
                          monotonicity_cutoff: float = 0.1):
    warnings = {
        "short_run": False,
        "large_spikes_in_train": np.nan,
        "large_deviation_in_train": np.nan,
        "low_end_in_train": np.nan,
        "non_monotonic": np.nan,
        "backtracking": np.nan,
        "runaway_test": np.nan,
        "non_convergence": np.nan
    }

    if 'learning_curve_train_elbo' not in adata.uns.keys():
        return warnings

    if adata.uns['learning_curve_train_epoch'][-1] < 50:
        warnings["short_run"] = True
        return warnings

    train_elbo_min_max = np.percentile(adata.uns['learning_curve_train_elbo'], q=[5, 95])
    train_elbo_range = train_elbo_min_max.max() - train_elbo_min_max.min()

    # look only from epoch 45 onward for spikes in train ELBO
    warnings["large_spikes_in_train"] = np.any((adata.uns['learning_curve_train_elbo'][46:]
                                                - adata.uns['learning_curve_train_elbo'][45:-1])
                                               < -train_elbo_range * spike_size)

    second_half_train_elbo = (adata.uns['learning_curve_train_elbo']
                              [(len(adata.uns['learning_curve_train_elbo']) // 2):])
    warnings["large_deviation_in_train"] = np.any(second_half_train_elbo
                                                  < np.median(second_half_train_elbo)
                                                  - train_elbo_range * deviation_size)

    half = len(adata.uns['learning_curve_train_elbo']) // 2
    threequarter = len(adata.uns['learning_curve_train_elbo']) * 3 // 4
    typical_end_variation = np.std(adata.uns['learning_curve_train_elbo'][half:threequarter])
    warnings["low_end_in_train"] = (adata.uns['learning_curve_train_elbo'][-1]
                                    < adata.uns['learning_curve_train_elbo'].max() - 5 * typical_end_variation)

    # look only from epoch 45 onward for spikes in train ELBO
    non_monotonicity = ((adata.uns['learning_curve_train_elbo'][46:]
                         - adata.uns['learning_curve_train_elbo'][45:-1])
                        < -3 * typical_end_variation).sum() / len(adata.uns['learning_curve_train_elbo'])
    warnings["non_monotonic"] = (non_monotonicity > monotonicity_cutoff)

    def windowed_cumsum(x, n=20):
        return np.array([np.cumsum(x[i:(i + n)])[-1] for i in range(len(x) - n)])

    windowsize = 20
    tracking_trace = windowed_cumsum(adata.uns['learning_curve_train_elbo'][1:]
                                     - adata.uns['learning_curve_train_elbo'][:-1],
                                     n=windowsize)
    big_dip = -1 * (adata.uns['learning_curve_train_elbo'][-1]
                    - adata.uns['learning_curve_train_elbo'][5]) / 10
    warnings["backtracking"] = (tracking_trace.min() < big_dip)

    halftest = len(adata.uns['learning_curve_test_elbo']) // 2
    threequartertest = len(adata.uns['learning_curve_test_elbo']) * 3 // 4
    typical_end_variation_test = np.std(adata.uns['learning_curve_test_elbo'][halftest:threequartertest])
    warnings["runaway_test"] = (adata.uns['learning_curve_test_elbo'][-1]
                                < adata.uns['learning_curve_test_elbo'].max() - 4 * typical_end_variation_test)

    warnings["non_convergence"] = (np.mean([adata.uns['learning_curve_train_elbo'][-1]
                                            - adata.uns['learning_curve_train_elbo'][-2],
                                            adata.uns['learning_curve_train_elbo'][-2]
                                            - adata.uns['learning_curve_train_elbo'][-3]])
                                   > 2 * typical_end_variation)

    return warnings


def plot_table(ax, values, colors=None, title=""):
    ax.axis('off')
    ax.axis('tight')
    df = pd.DataFrame(values, index=[0]).T
    max_key_length = max([len(str(key)) for key in df.index])
    max_value_length = max([len(str(value)) for value in df[0]])
    total_length = max_key_length + max_value_length
    table = ax.table(cellText=df.values, colWidths=[total_length / max_key_length, total_length / max_value_length], rowLabels=df.index, loc='center', edges='open')
    table.auto_set_column_width(col=list(range(len(df.columns))))
    table.scale(1, 0.8)
    if colors is not None:
        for row_index in range(len(values)):
            key = table[(row_index, -1)].get_text()
            key.set_color(colors[key.get_text()])
            value = table[(row_index, 0)].get_text()
            value.set_color(colors[key.get_text()])
    ax.set_title(title)


def plot_train_error(adata, ax):
    if 'learning_curve_train_elbo' not in adata.uns.keys():
        print('No learning curve recorded!')
        ax.set_axis_off()

    try:
        ax.plot(adata.uns['learning_curve_train_epoch'],
                adata.uns['learning_curve_train_elbo'], '.--', label='Train')
        try:
            ax.plot(adata.uns['learning_curve_test_epoch'],
                    adata.uns['learning_curve_test_elbo'], 'o:', label='Test')
            ax.legend()
        except Exception:
            pass

        ylim_low = max(adata.uns['learning_curve_train_elbo'][0], adata.uns['learning_curve_train_elbo'][-1] - 2000)
        try:
            ylim_high = max(max(adata.uns['learning_curve_train_elbo']), max(adata.uns['learning_curve_test_elbo']))
        except ValueError:
            ylim_high = max(adata.uns['learning_curve_train_elbo'])
        ylim_high = ylim_high + (ylim_high - ylim_low) / 20
        ax.set_ylim([ylim_low, ylim_high])
    except:
        print('Error plotting the learning curve. Skipping.')
        pass

    ax.set_xlabel('Epoch')
    ax.set_ylabel('ELBO')
    ax.set_title('Progress of the training procedure')


def plot_barcodes_and_inferred_cell_prop(adata, ax):
    in_counts = np.array(adata.layers['raw'][:, adata.var['cellbender_analyzed']].sum(axis=1)).squeeze()
    order = np.argsort(in_counts)[::-1]
    ax.semilogy(in_counts[order], color='black')
    ax.set_ylabel('UMI counts')
    ax.set_xlabel('Barcode index, sorted by UMI count')
    right_ax = ax.twinx()
    right_ax.plot(adata.obs['cell_probability'][order].values, '.:', color='red', alpha=0.3, rasterized=True)
    right_ax.set_ylabel('Cell probability', color='red')
    right_ax.set_ylim([-0.05, 1.05])
    ax.set_title('Determination of which barcodes contain cells')


def plot_latent_encoding_z(adata, ax):
    cells = (adata.obs['cell_probability'] > 0.5)
    adata.obsm['X_pca'] = pca_2d(adata.obsm['cellbender_embedding']).detach().numpy()

    # Plot the latent encoding via PCA.
    ax.plot(adata.obsm['X_pca'][:, 0][cells],
            adata.obsm['X_pca'][:, 1][cells],
            '.',
            ms=3,
            color='black',
            alpha=0.3,
            rasterized=True)
    ax.set_ylabel('PC 1')
    ax.set_xlabel('PC 0')
    ax.set_title('PCA of latent encoding of gene expression in cells')


def pca_2d(mat: np.ndarray) -> torch.Tensor:
    A = torch.as_tensor(mat).float()
    U, S, V = torch.pca_lowrank(A)
    return torch.matmul(A, V[:, :2])

def plot_report(input_files, suffix="1"):
    nrows = len(input_files)

    fig, axs = plt.subplots(nrows, 5, figsize=(30, 6 * nrows), gridspec_kw={"width_ratios": [0.05, 0.05, 0.3, 0.3, 0.3]})
    if nrows == 1:
        axs = axs[np.newaxis, ...]

    for row_index, (sample, input_h5_path, settings_path, raw_h5_path) in enumerate(input_files):
        print("\tRow {} - sample: {}:".format(row_index, sample))
        print("\t  --input_h5 {}".format(input_h5_path))
        print("\t  --settings {}".format(settings_path))
        print("\t  --raw_h5 {}".format(raw_h5_path))
        print("")

        fh = open(settings_path)
        settings = {}
        colors = {}
        for key, value in json.load(fh).items():
            if key in settings_info:
                new_key, default_value = settings_info[key]
                settings[new_key] = str(value)
                if str(value) == str(default_value):
                    colors[new_key] = "black"
                else:
                    colors[new_key] = "red"
        fh.close()
        plot_table(
            ax=axs[row_index, 0],
            values=settings,
            colors=colors,
            title=sample
        )

        # load datasets, before and after CellBender
        adata = load_anndata_from_input_and_output(input_file=input_h5_path,
                                                   output_file=raw_h5_path,
                                                   analyzed_barcodes_only=True,
                                                   input_layer_key='raw',
                                                   truth_file=None)
        raw_full_adata = load_anndata_from_input(input_h5_path)

        # bit of pre-compute
        adata.var['n_removed'] = adata.var[f'n_raw'] - adata.var[f'n_cellbender']

        warnings = {}
        try:
            warnings.update(assess_overall_count_removal(adata))
        except ValueError:
            print('Skipping assessment over overall count removal. Presumably '
                  'this is due to including the whole dataset in '
                  '--total-droplets-included.')

        try:
            warnings.update(assess_learning_curve(adata))
        except Exception:
            pass

        # look at per-gene count removal
        warnings.update(assess_count_removal_per_gene(adata, raw_full_adata=raw_full_adata))
        has_warnings = False
        colors = {}
        for key, value in warnings.items():
            if value is True or value == "incorrect" or value == "more" or value == "fewer":
                has_warnings = True
                colors[key] = "red"
            else:
                colors[key] = "green"
        warnings["--------"] = "--------"
        warnings["Warnings"] = has_warnings
        colors["--------"] = "black"
        colors["Warnings"] = "red" if has_warnings else "green"

        plot_table(
            ax=axs[row_index, 1],
            values=warnings,
            colors=colors,
            title="Warnings"
        )
        plot_train_error(
            adata=adata,
            ax=axs[row_index, 2]
        )
        plot_barcodes_and_inferred_cell_prop(
            adata=adata,
            ax=axs[row_index, 3]
        )
        plot_latent_encoding_z(
            adata=adata,
            ax=axs[row_index, 4]
        )

    fig.tight_layout()
    outfile = os.path.join(args.out, 'CellBender_report.{}.png'.format(suffix))
    plt.savefig(outfile, bbox_inches="tight")
    return outfile


################################################################################

input_files = list(zip(args.samples, args.input_h5, args.settings, args.raw_h5))
input_files.sort()

nplots = len(input_files)

start_indices = [start_index for start_index in range(0, nplots, args.max_plots_per_page)]
n_figures = len(start_indices)
start_indices.append(start_indices[-1] + args.max_plots_per_page)

print("Plotting")
for i in range(n_figures):
    reportfile = plot_report(
        input_files=input_files[start_indices[i]:start_indices[i + 1]],
        suffix=str(i)
    )
    print("\tSaved {}".format(reportfile))
    print("")


scripts/visualise_cellbender_results.py
#!/usr/bin/env python3

"""
File:         visualise_cellbender_results.py
Created:      2023/07/07
Last Changed: 2023/09/13
Author:       M.Vochteloo
"""

# Standard imports.
from __future__ import print_function
from datetime import datetime
import argparse
import pathlib
import math
import os
import re

# Third party imports.
import numpy as np
import pandas as pd
import h5py
import torch
import seaborn as sns
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

# Local application imports.

# Metadata
__program__ = "Visualise CellBender Results"
__author__ = "Martijn Vochteloo"
__maintainer__ = "Martijn Vochteloo"
__email__ = "m.vochteloo@rug.nl"
__license__ = "GPLv3"
__version__ = 1.0
__description__ = "{} is a program developed and maintained by {}. " \
                  "This program is licensed under the {} license and is " \
                  "provided 'as-is' without any warranty or indemnification " \
                  "of any kind.".format(__program__,
                                        __author__,
                                        __license__)

parser = argparse.ArgumentParser(prog=__program__,
                                 description=__description__)

# Add other arguments.
parser.add_argument("--pools", nargs="*", type=str, required=True, help="")
parser.add_argument("--raw_h5", nargs="*", type=str, required=True, help="")
parser.add_argument("--posterior_h5", nargs="*", type=str, required=True, help="")
parser.add_argument("--metrics", nargs="*", type=str, required=True, help="")
parser.add_argument("--log", nargs="*", type=str, required=True, help="")
parser.add_argument("--output", type=str, required=True, help="")
args = parser.parse_args()


def parse_log_file(logfile):
    data = {}

    start_datetime = None
    end_datetime = None
    completed = False
    with open(logfile, 'r') as f:
        for line in f:
            if line.startswith("cellbender remove-background"):
                for part in line.split("--")[1:]:
                    if "=" in part:
                        key, value = part.split("=")
                        try:
                            data["parameter:" + key] = float(value.strip("\n"))
                        except ValueError:
                            pass
                    else:
                        data[part] = 1
            elif "Completed remove-background." in line:
                data["passed"] = True
            elif re.search("(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})", line):
                tmp_datetime = datetime.strptime(re.search("(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})", line).group(0), "%Y-%m-%d %H:%M:%S")
                if start_datetime is None:
                    start_datetime = tmp_datetime
                elif completed:
                    end_datetime = tmp_datetime
            elif re.search("Features in dataset: ([0-9]+) Gene Expression", line):
                data["features"] = int(re.search("Features in dataset: ([0-9]+) Gene Expression", line).group(1))
            elif re.search("([0-9]+) features have nonzero counts.", line):
                data["nonzero_genes"] = int(re.search("([0-9]+) features have nonzero counts.", line).group(1))
            elif re.search("Prior on counts for cells is ([0-9]+)", line):
                data["prior_counts_empty"] = int(re.search("Prior on counts for cells is ([0-9]+)", line).group(1))
            elif re.search("Prior on counts for empty droplets is ([0-9]+)", line):
                data["prior_counts_cell"] = int(re.search("Prior on counts for empty droplets is ([0-9]+)", line).group(1))
            elif re.search("Excluding ([0-9]+) features that are estimated to have <= 0.1 background counts in cells.", line):
                data["excluded_features"] = int(re.search("Excluding ([0-9]+) features that are estimated to have <= 0.1 background counts in cells.", line).group(1))
            elif re.search("Including ([0-9]+) features in the analysis.", line):
                data["included_features"] = int(re.search("Including ([0-9]+) features in the analysis.", line).group(1))
            elif re.search("Excluding barcodes with counts below ([0-9]+)", line):
                data["barcodes_threshold"] = int(re.search("Excluding barcodes with counts below ([0-9]+)", line).group(1))
            elif re.search("Using ([0-9]+) probable cell barcodes, plus an additional ([0-9]+) barcodes, and ([0-9]+) empty droplets.", line):
                match = re.search("Using ([0-9]+) probable cell barcodes, plus an additional ([0-9]+) barcodes, and ([0-9]+) empty droplets.", line)
                data["probable_cell_barcodes"] = int(match.group(1))
                data["additional_barcodes"] = int(match.group(2))
                data["empty_droplets"] = int(match.group(3))
            elif re.search("Largest surely-empty droplet has ([0-9]+) UMI counts.", line):
                data["max_empty_droplet_count"] = int(re.search("Largest surely-empty droplet has ([0-9]+) UMI counts.", line).group(1))
            elif "Learning failed.  Retrying with learning-rate " in line:
                data = {}
                start_datetime = end_datetime
                end_datetime = None
            elif "Completed remove-background." in line:
                completed = True
            else:
                pass
    f.close()

    if start_datetime is not None and end_datetime is not None:
        data["time"] = (end_datetime - start_datetime).total_seconds() / 60.0

    return data


def barplot(df, panels=None, x="x", y="y", panel_column=None,
            palette=None, order=None, title="", filename="plot"):
    if panels is None:
        panels = df[panel_column].unique()
    if order is None:
        order = df[x].unique()
        order.sort()

    nplots = len(panels)
    ncols = math.ceil(np.sqrt(nplots))
    nrows = math.ceil(nplots / ncols)

    sns.set_style("ticks")
    fig, axes = plt.subplots(nrows=nrows,
                             ncols=ncols,
                             sharex='all',
                             sharey='none',
                             figsize=(6 * ncols, 6 * nrows))
    sns.set(color_codes=True)

    row_index = 0
    col_index = 0
    for i in range(ncols * nrows):
        if nrows == 1 and ncols == 1:
            ax = axes
        elif nrows == 1 and ncols > 1:
            ax = axes[col_index]
        elif nrows > 1 and ncols == 1:
            ax = axes[row_index]
        else:
            ax = axes[row_index, col_index]

        if i < nplots:
            data = df.loc[df[panel_column] == panels[i], :]
            if data.shape[0] == 0:
                continue

            sns.barplot(
                data=data,
                x=x,
                y=y,
                color="black",
                palette=palette,
                order=order,
                ax=ax
            )

            ax.set_xticklabels(ax.get_xmajorticklabels(), rotation=90)
            ax.set_ylim(int(data[y].min() - (data[y].max() * 0.05)), ax.get_ylim()[1])

            ax.set_xlabel("",
                          fontsize=10,
                          fontweight='bold')
            ax.set_ylabel("",
                          fontsize=10,
                          fontweight='bold')
            ax.set_title(panels[i],
                         fontsize=14,
                         fontweight='bold')

        else:
            ax.set_xticklabels(ax.get_xmajorticklabels(), rotation=90)

        col_index += 1
        if col_index > (ncols - 1):
            col_index = 0
            row_index += 1

    fig.suptitle(title,
                 fontsize=40,
                 fontweight='bold')

    outpath = os.path.join(args.output, filename)
    fig.savefig(outpath)
    print("\tSaved figure: {}".format(os.path.basename(outpath)))
    plt.close()


def plot_per_sample(df, plottype, samples=None, subtitles=None, x="x", y="y", y2=None,
                    alpha=1.0, sample_column="sample", hue=None, color="black", palette=None,
                    xlabel="", ylabel="", ylabel2="", title="", filename="plot"):
    if samples is None:
        samples = df[sample_column].values.tolist()
        samples.sort()
    if subtitles is not None and len(subtitles) != len(samples):
        print("Error, subtitles are not the same length as the samples")
        return

    nplots = len(samples)
    ncols = math.ceil(np.sqrt(nplots))
    nrows = math.ceil(nplots / ncols)

    sns.set_style("ticks")
    fig, axes = plt.subplots(nrows=nrows,
                             ncols=ncols,
                             sharex='none',
                             sharey='none',
                             figsize=(6 * ncols, 6 * nrows))
    sns.set(color_codes=True)

    row_index = 0
    col_index = 0
    for i in range(ncols * nrows):
        if nrows == 1 and ncols == 1:
            ax = axes
        elif nrows == 1 and ncols > 1:
            ax = axes[col_index]
        elif nrows > 1 and ncols == 1:
            ax = axes[row_index]
        else:
            ax = axes[row_index, col_index]

        if i < nplots:
            data = df.loc[df[sample_column] == samples[i], :].dropna()
            if data.shape[0] == 0:
                continue

            subtitle = ""
            if subtitles is not None:
                subtitle = subtitles[i]

            sns.despine(fig=fig, ax=ax)

            if plottype == "line":
                sns.lineplot(data=data,
                             x=x,
                             y=y,
                             hue=hue,
                             style=hue,
                             color=color,
                             alpha=alpha,
                             markers=["o"] * len(data[hue].unique()) if hue in data else "o",
                             markeredgewidth=0.0,
                             palette=palette,
                             ax=ax)
            elif plottype == "scatter":
                sns.scatterplot(data=data,
                                x=x,
                                y=y,
                                hue=hue,
                                color=color,
                                alpha=alpha,
                                s=10,
                                palette=palette,
                                ax=ax)
            elif plottype == "multi":
                g = sns.lineplot(data=data,
                                 x=x,
                                 y=y,
                                 hue=hue,
                                 color="black",
                                 alpha=1,
                                 palette=palette,
                                 ax=ax)
                ax.set(yscale="log")
                ax2 = ax.twinx()

                sns.scatterplot(data=data,
                                x=x,
                                y=y2,
                                hue=hue,
                                color=color,
                                alpha=alpha,
                                s=10,
                                palette=palette,
                                ax=ax2)

                ax2.set_ylabel(ylabel2,
                               color=color,
                               fontsize=10,
                               fontweight='bold')

            ax.set_xlabel(xlabel,
                          fontsize=10,
                          fontweight='bold')
            ax.set_ylabel(ylabel,
                          fontsize=10,
                          fontweight='bold')
            ax.set_title("{}{}".format(samples[i], subtitle),
                         fontsize=14,
                         fontweight='bold')


        else:
            ax.set_axis_off()

        col_index += 1
        if col_index > (ncols - 1):
            col_index = 0
            row_index += 1

    fig.suptitle(title,
                 fontsize=16,
                 color="#000000",
                 weight='bold')

    outpath = os.path.join(args.output, filename)
    fig.savefig(outpath)
    print("\tSaved figure: {}".format(os.path.basename(outpath)))
    plt.close()


#############################
############ STATS ##########
#############################
stats_data = []
for pool, metrics_path, log_path in zip(args.pools, args.metrics, args.log):
    # Load the metrics.
    metrics_df = pd.read_csv(metrics_path, sep=",")
    metrics_df.columns = ["key", "value"]
    metrics_df["value"] = metrics_df["value"].astype(float)
    metrics_df["id"] = pool + "_" + pathlib.PurePath(metrics_path).parent.name

    # Add the metrics from the log file.
    log_df = pd.DataFrame(parse_log_file(log_path), index=[0]).T.reset_index(drop=False)
    log_df.columns = ["key", "value"]
    metrics_df["id"] = pool + "_" + pathlib.PurePath(log_path).parent.name

    # Save the data.
    stats_data.append(metrics_df)
    stats_data.append(log_df)

stats_df = pd.concat(stats_data, axis=1)
barplot(
    df=stats_df,
    x="id",
    y="value",
    panel_column="key",
    filename="stats.png"
)
del stats_data, stats_df

##########################################
############ TRAINING PROCEDURE ##########
##########################################
training_procedure_data = []
for pool, raw_h5_path in zip(args.pools, args.raw_h5):
    hf = h5py.File(raw_h5_path, 'r')
    training_procedure_data.append(pd.DataFrame({
        "epoch": np.array(hf.get('metadata/learning_curve_train_epoch')),
        "loss": np.array(hf.get('metadata/learning_curve_train_elbo')),
        "id": pool + "_" + pathlib.PurePath(raw_h5_path).parent.name,
        "group": "Train"
    }))
    training_procedure_data.append(pd.DataFrame({
        "epoch": np.array(hf.get('metadata/learning_curve_test_elbo')),
        "loss": np.array(hf.get('metadata/learning_curve_test_epoch')),
        "id": pool + "_" + pathlib.PurePath(raw_h5_path).parent.name,
        "group": "Test"
    }))
    hf.close()

training_procedure_df = pd.concat(training_procedure_data, axis=1)
plot_per_sample(
    df=training_procedure_df,
    plottype="line",
    x="epoch",
    y="loss",
    hue="group",
    xlabel="Epoch",
    ylabel="ELBO",
    title="Progress of the training procedure",
    filename="training_procedure.png"
)
del training_procedure_data, training_procedure_df

########################################
############ CELL PROBABILITY ##########
########################################
cell_probability_data = []
for pool, posterior_h5_path in zip(args.pools, args.posterior_h5):
    hf = h5py.File(posterior_h5_path, 'r')
    cell_probability_data.append(pd.DataFrame({
        "d": np.array(hf.get('droplet_latents_map/d')),
        "p": np.array(hf.get('droplet_latents_map/p')),
        "id": pool + "_" + pathlib.PurePath(posterior_h5_path).parent.name
    }))
    hf.close()

cell_probability_df = pd.concat(cell_probability_data, axis=1)
# TODO: reorder based on cell size. I tried this but then the cell procentage scatterplot does not match the
#  CellBender output figure. If I don't reorder, then the UMI line does not match. I think I need to
#  order the UMI line but not the counts but that feels wrong.
plot_per_sample(
    df=cell_probability_df,
    plottype="multi",
    x="index",
    y="d",
    y2="p",
    color="red",
    alpha=0.3,
    xlabel="Barcode index",
    ylabel="UMI counts",
    ylabel2="Cell probability",
    title="Determination of which barcodes contain cells",
    filename="cell_probability.png"
)
del cell_probability_data, cell_probability_df

############################################
############ LATENT GENE ENCODING ##########
############################################
latent_gene_encoding_data = []
for pool, posterior_h5_path in zip(args.pools, args.posterior_h5):
    hf = h5py.File(posterior_h5_path, 'r')

    # Extract the data.
    p = np.array(hf.get('droplet_latents_map/p'))
    z = np.array(hf.get('droplet_latents_map/z'))

    # Calculate PCA.
    A = torch.as_tensor(z[p >= 0.5]).float()
    U, S, V = torch.pca_lowrank(A)
    z_pca = torch.matmul(A, V[:, :2]).to_numpy()

    # Save data.
    latent_gene_encoding_data.append(pd.DataFrame({
        "PC0": z_pca[:, 0],
        "PC1": z_pca[:, 1],
        "id": pool + "_" + pathlib.PurePath(posterior_h5_path).parent.name
    }))
    hf.close()

latent_gene_encoding_df = pd.concat(latent_gene_encoding_data, axis=1)
plot_per_sample(
    df=latent_gene_encoding_df,
    plottype="scatter",
    x="PC0",
    y="PC1",
    alpha=0.3,
    xlabel="PC 0",
    ylabel="PC 1",
    title="PCA of latent encoding of cell gene expression",
    filename="latent_gene_encoding"
)
del latent_gene_encoding_data, latent_gene_encoding_df
